File: G_Projects/G_Inventory_Management_System/UtilityClasses/Warehouse.java
package G_Projects.G_Inventory_Management_System.UtilityClasses;

import java.util.Collection;
import java.util.HashMap;
import java.util.Map;

import G_Projects.G_Inventory_Management_System.ProductFactory.Product;

public class Warehouse {
    private String name;
    private String location;
    
    private Map<String, Product> products;

    public Warehouse(String name){
        this.name = name;
        this.products = new HashMap<>();
    }

    public void setLocation(String location){
        this.location = location;
    }

    public void addProduct(Product product, int quantity){
        String id = product.getId();
        if(products.containsKey(id)){
            // Product already exists, update quantity
            Product existingProduct = products.get(id);
            existingProduct.setQuantity(existingProduct.getQuantity() + quantity);
        }else{
            product.setQuantity(quantity); 
        }
        products.put(id, product);
        System.out.println("Added " + quantity + " of product ID: " + id + " to warehouse: " + name + ". New quantity: " + getAvailableQuantity(id));
    }

    // remove product from warehouse
    public boolean removeProduct(String id, int quantity){
        if(products.containsKey(id)){
            Product product = products.get(id);
            int currQuantity = product.getQuantity();
            if(currQuantity >= quantity){
                // safe to remove
                product.setQuantity(currQuantity - quantity);
                System.out.println("Removed " + quantity + " of product ID: " + id + " from warehouse: " + name + ". Remaining Quantity: " + product.getQuantity());

                if(product.getQuantity() == 0){
                    products.remove(id);
                    System.out.println("Product ID: " + id + " is out of stock and removed from warehouse: " + name);
                }
                return true;
            }else{
                System.out.println("Insufficient stock for product ID: " + id + " in warehouse: " + name + ". Available: " + currQuantity + ", Requested: " + quantity);
                return false;
            }
        }else{
            System.out.println("Product ID: " + id + " not found in warehouse: " + name);
            return false;
        }
    }

    public int getAvailableQuantity(String id){
        if(products.containsKey(id)){
            return products.get(id).getQuantity();
        }
        return 0;
    }

    public Product getProductByID(String id){
        return products.getOrDefault(id, null);
    }

    public Collection<Product> getAllProducts(){
        return products.values();
    }
}

File: Python/A_Project/D_Inventory_Management_System/UtilityClasses/warehouse.py
class Warehouse:
    def __init__(self, name: str):
        self.name = name
        self.location = None
        self.products = {}

    def set_location(self, location: str):
        self.location = location

    def add_product(self, product, quantity: int):
        product_id = product.get_id()
        if product_id in self.products:
            existing_product = self.products[product_id]
            existing_product.set_quantity(existing_product.get_quantity() + quantity)
        else:
            product.set_quantity(quantity)
        self.products[product_id] = product
        print(
            f"Added {quantity} of product ID: {product_id} to warehouse: {self.name}. "
            f"New quantity: {self.get_available_quantity(product_id)}"
        )

    def remove_product(self, product_id: str, quantity: int) -> bool:
        if product_id in self.products:
            product = self.products[product_id]
            current_quantity = product.get_quantity()
            if current_quantity >= quantity:
                product.set_quantity(current_quantity - quantity)
                print(
                    f"Removed {quantity} of product ID: {product_id} from warehouse: "
                    f"{self.name}. Remaining Quantity: {product.get_quantity()}"
                )
                if product.get_quantity() == 0:
                    self.products.pop(product_id, None)
                    print(
                        f"Product ID: {product_id} is out of stock and removed from "
                        f"warehouse: {self.name}"
                    )
                return True
            print(
                f"Insufficient stock for product ID: {product_id} in warehouse: "
                f"{self.name}. Available: {current_quantity}, Requested: {quantity}"
            )
            return False
        print(f"Product ID: {product_id} not found in warehouse: {self.name}")
        return False

    def get_available_quantity(self, product_id: str) -> int:
        if product_id in self.products:
            return self.products[product_id].get_quantity()
        return 0

    def get_product_by_id(self, product_id: str):
        return self.products.get(product_id)

    def get_all_products(self):
        return self.products.values()
